rework to support more sorts such as bitvector

archaistvolts opened the pull request:
the current design doesn't support mk_*_sort methods with more than one arg.  this patch attempts to solve that by roughly following the structures from prove-rs/z3.rs.

TODO: add back support for `optimize` and various methods such as implies(), ite(), etc.

* add Context type
	* add errorHandler which prints output from `Z3_get_error_msg()`
* add tests.zig with passing tests "solve for model" and "bitvectors"
* update example.basic() to new API
* install artifacts and enable valgrind in build.zig so we can use valgrind.  not sure if this has much value since z3 seems to report some leaks on its own.

archaistvolts committed:
rework to support more sorts such as bitvector

the current design doesn't support mk_*_sort methods with more than one
arg.  this patch attempts to solve that by roughly following the
structures from prove-rs/z3.rs.

TODO: add back support for `optimize` and various methods such as
implies(), ite(), etc.

* add Context type
	* add errorHandler which prints output from `Z3_get_error_msg()`
* add tests.zig with passing tests "solve for model" and "bitvectors"
* update example.basic() to new API
* install artifacts and enable valgrind in build.zig so we can use
  valgrind.  not sure if this has much value since z3 seems to report
  some leaks on its own.

archaistvolts commented:
I've marked as draft because I want to add back support for things I've removed.  And I'm looking for early feedback to get an idea whether you think this is a good approach.

I know the builder pattern isn't usually used in zig codebases.  But so far the only downsides are copying Context in each Ast struct and having to copy the `Z3_ast`s out to a temp buffer in methods which allow more than one rhs param such as add/sub/mul.

I'll add more commits soon which add back things I've removed.

archaistvolts commented on src/lib.zig:
```diff
@@ -1,264 +1,476 @@
+            if (buf.len < rhss.len + 1) @panic("numericVarop only supports up to 15 rhs args.");
+            buf[0] = lhs.ast;
+            for (0..rhss.len) |i| buf[i + 1] = rhss[i].ast;
+            const ast = func(lhs.ctx.inner, @intCast(rhss.len + 1), &buf);
```
I've assumed that `buf` is being duped by z3 here.  I don't know for sure if this is true.  But so far this doesn't seem to cause any problems.

Rexicon226 commented:
Thanks for the PR! I will take a look at it later today and give some initial feedback. I just pushed some commits that update this project to 0.15, and it doesn't seem to have created any conflicts with your branch! Just wanted to let you know.

archaistvolts committed:
initial support for optimize with test

* rename Solver -> Model, Model -> PartialModel

Rexicon226 commented on src/lib.zig:
```diff
@@ -1,264 +1,507 @@
+        fn numericVarop(lhs: Self, rhss: []const Self, comptime func: @TypeOf(c.Z3_mk_add)) Self {
+            comptime if (!T.ast_kind.isNumeric()) unreachable;
+            var buf: [16]c.Z3_ast = undefined;
+            if (buf.len < rhss.len + 1) @panic("numericVarop only supports up to 15 rhs args.");
```
This can be a compile error instead of a panic.

Rexicon226 commented on src/lib.zig:
```diff
@@ -1,264 +1,507 @@
+        }
+        const inner = c.Z3_mk_context_rc(cfg);
+        c.Z3_set_error_handler(inner, errorHandler);
+        const cached_sorts = try allocator.create(CachedSorts);
```
I wonder if we could just keep the sorts inline, and store a reference to `Context` within the asts. The user can then decide whether to just create the whole `Context` on the heap themselves and then initialize it or store it in a higher frame if it might go out of scope.
I don't really like the idea of forcing an allocator here, it doesn't seem necessary.

archaistvolts replied:
I've attempted to address this.  Re-reading now and i'm not 100% sure i've followed correctly since Model now stores Context by value while everywhere else stores *Context.  But that seems fine to me since were always storing (and all methods accept) either `*const Context` or `*Context`.

Rexicon226 commented on src/lib.zig:
```diff
@@ -1,264 +1,507 @@
-};
+        // *** Numeric ops ***
+        fn numericBinop(R: type, lhs: Self, rhs: Self, func: @TypeOf(c.Z3_mk_div)) R {
+            comptime if (!T.ast_kind.isNumeric()) unreachable;
```
Could we make these `@compileError` to give more useful error messages?

Rexicon226 commented on src/tests.zig:
```diff
@@ -0,0 +1,86 @@
+    var optimize: z3.Model = try .initConfig(
+        .optimize,
+        gpa,
+        &.{.{ "timeout", "1" }}, // 1 ms timeout
```
I don't really like this, there is a well-defined list of configs. This should be a config struct that is serialized to strings or calls the config functions directly.
https://github.com/Z3Prover/z3/blob/49703f8bba0e73fbd2aa6b180f8afdaeadd4d7a4/src/util/gparams.cpp#L44-L73
https://github.com/Z3Prover/z3/blob/49703f8bba0e73fbd2aa6b180f8afdaeadd4d7a4/src/params/context_params.cpp#L63-L132

archaistvolts replied:
I've created a ConfigParam type.  I've left a TODO to support renames and some other keys i haven't found types for yet.

Rexicon226 commented on src/lib.zig:
```diff
@@ -1,264 +1,507 @@
+        }
+
+        /// Addition
+        pub fn bvadd(lhs: Bitvector, rhs: Bitvector) Bitvector {
```
Hmm, I understand why you took this approach with putting all of the functions for each sort into the same namespace, but I feel there is a better way to achieve this. If you want to keep the ability to do member function calls, `anytype` can get you pretty far.

```zig
const A = struct {
	fn add(x: anytype, y: @TypeOf(x)) @TypeOf(x) {
		...
	}
};

const B = struct {
	const add = A.add;
}
...
var x: B = ...;
x.add(y);
```
you'll find that this works (if i remember correctly). I feel that this approach can be applied here. Have common `add` implementations that you import into each sort with `const add = generic.add;` and for things like bv we have custom implementations.

What do you think about that?

Rexicon226 commented on src/lib.zig:
```diff
@@ -1,264 +1,507 @@
-    pub fn le(m: *Model, lhs: anytype, rhs: anytype) Bool {
-        comptime std.debug.assert(@TypeOf(lhs) == @TypeOf(rhs));
-        return .{ .ast = z3.Z3_mk_le(m.ctx, lhs.ast, rhs.ast) };
+    pub fn double(m: Model) Sort {
```
Let's call this `float64`, or even better, just make it take a float type as an argument.

Also consider using `std.math.floatExponentBits(` and `std.math.floatMantissaBits() + 1`.

Rexicon226 reviewed (changes requested):
Gave it an initial run through, overall I like the ideas, but there are some things I'd like to change.

archaistvolts committed:
address initial review comments

* store cached sorts by value and use *Context in asts
* remove allocator param from Model/Context init* methods.  Context now
  stores cached_sorts by value, Model has Context, PartialModel stores
  *const Context.
* @compileError instead of comptime unreachables
* start using anytype approach with decl pasting instead of keeping all
  methods in a single namespace
* Model.float() now accepts a type

archaistvolts commented:
Thanks for the reviews!  I've pushed a commit addressing them and marked most comments as resolved for now.

archaistvolts committed:
add test "dynamic as set" with initial array support

archaistvolts committed:
add back things i removed such as implies(), ite(), etc

archaistvolts committed:
add test "ite"

archaistvolts committed:
add test "array example 1"

* remove AstKind for SortKind
* Ast is now a regular struct instead of a type fn

archaistvolts committed:
whitespace and rename only to make prev commit readable

archaistvolts commented:
I'll stop here for now and give you a chance to review.  I'd like to add more test coverage as I'm sure some code isn't covered.  But I feel like this is might be a good place to stop for now.  Hope its not too much.

archaistvolts commented on build.zig:
```diff
@@ -101,9 +101,23 @@ pub fn build(b: *std.Build) !void {
     example.root_module.addImport("z3", z3_bindings);
-
+    // enable running `valgrind --leak-check=full zig-out/bin/main`
+    example.root_module.valgrind = true;
```
Should `enable-valgrind` be a build option?  I'm actually not entirely sure what this does.  I assume it instruments the exe so that zig allocators can be detected/used by valgrind?  If thats true then I guess it should be removed since we're no longer using zig allocators.

archaistvolts replied:
Went ahead and removed the valgrind stuff.

archaistvolts committed:
remove non-common methods from Ast to Bool/Bitvector

* build: remove valgrind since we're not using zig allocators

archaistvolts committed:
bring back AstKind enum as Ast.Tag

* move SortKind enum to Sort.Tag

archaistvolts committed:
handle ConfigParam enum types

Rexicon226 commented:
Sorry for the delay, I'll give this a second round of reviews soon (within like 5 hours), have been a bit tied down with work :).

archaistvolts commented:
Sounds good.  Timing seems good.  Glad I've had time to review and fix a few things.

Rexicon226 commented on build.zig:
```diff
@@ -101,9 +101,20 @@ pub fn build(b: *std.Build) !void {
+        .imports = &.{.{ .name = "z3", .module = z3_bindings }},
+    }) });
+    b.installArtifact(tests);
+    tests.filters = if (b.option([]const u8, "test-filter", "test filter")) |f| b.dupeStrings(&.{f}) else &.{};
```
```suggestion
    const tests = b.addTest(.{ .root_module = b.createModule(.{
        .root_source_file = b.path("src/tests.zig"),
        .optimize = optimize,
        .target = target,
        .imports = &.{.{ .name = "z3", .module = z3_bindings }},
        .filters = b.option([]const []const u8), "test-filter", "") orelse &.{},
    }) });
    b.installArtifact(tests);
```

or something like that

Rexicon226 commented on src/lib.zig:
```diff
@@ -1,264 +1,741 @@
+    };
+
+    pub const ConfigParam = union(enum) {
+        timeout: c_uint,
```
these types are specifically for translate-c, use `u32` instead.

Rexicon226 commented on src/lib.zig:
```diff
@@ -1,264 +1,741 @@
+    pub fn float(m: *Model, comptime T: type) Sort {
+        const ebits: c_uint = std.math.floatExponentBits(T);
+        const sbits: c_uint = std.math.floatMantissaBits(T) + 1;
+        if (@typeInfo(T) != .float) @compileError("expected floating point type. found '" ++ @typeName(T) ++ "'");
```
not really useful, cuz `floatExponentBits` would have already failed if `T` wasn't a float.

Rexicon226 commented on src/lib.zig:
```diff
@@ -1,264 +1,741 @@
+        if (@typeInfo(T) != .float) @compileError("expected floating point type. found '" ++ @typeName(T) ++ "'");
+        switch (T) {
+            f32 => {
+                comptime std.debug.assert(ebits == 8 and sbits == 24);
```
I don't think we need to assert this, I am 100% certain that it will be the correct numbers.

Something like,
```zig
fn float(m: *Model, comptime T: type) Sort {
	const ebits = std.math.floatExponentBits(T);
	const sbits = std.math.floatMantissaBits(T) + 1;
	const tag = switch (T) {
		f32 => .float32,
		f64 => .float64,
		else => .float,
	};
	return m.ctx.getSortByName(tag, "Z3_mk_fpa_sort", .{ ebits, sbits });
}
```
(wrote by hand, haven't tried to compile).

Rexicon226 commented on src/lib.zig:
```diff
@@ -1,264 +1,741 @@
+        // TODO support additional params and renames: https://github.com/Z3Prover/z3/blob/49703f8bba0e73fbd2aa6b180f8afdaeadd4d7a4/src/util/gparams.cpp#L44-L73
+    };
+
+    pub const Config = []const ConfigParam;
```
Having it be a slice leads to the ability to have duplicate configs. Maybe we could make it a struct with optional fields? Just a thought, maybe there's some reasoning/usecase I don't see :).

Rexicon226 reviewed (changes requested):
It all looks very good, just a note on the style, would you mind putting a new line between functions?

archaistvolts committed:
add missing newlines between decls

* remove 'pub' from all Ast decls.

archaistvolts committed:
address 2nd review comments

* build filters
* change Config to struct of optionals with default null fields
* c_uint -> u32

archaistvolts committed:
add test "float sizes"

* Ast: add getSort()
* Sort: add kind(), exponentSize(), significandSize()
* Model: add constantFloat()

archaistvolts commented:
Thanks for the reviews!  I think I've addressed them all.

I've also added one more test to cover Model.float() with a few more supporting functions.  I'm not 100% sure about the name/API of `Model.constantWithSort()`.  But it seemed necessary for the new test "float sizes".

archaistvolts commented:
~~I've removed Model.constantWithSort() for Model.constantFloat(T).  It seems simpler now but I'm still not sure this is the best way.~~

archaistvolts committed:
Model.constant(): args use lib types instead of c types

I didn't like how Model.constant() with .array required to pass
`sort.inner`.  Now it calls helpers such as Model.int() and
Model.array() instead of calling Context.getSort().  This
changes it's usage from args `.{domain_sort.inner, range_sort.inner}` to
`.{domain_sort, range_sort}`.

* add required float helpers such as Model.float32() to match Ast.Tags
* remove constantFloat()
* add back constantWithSort()

archaistvolts commented:
Sorry for all the noise here.  I waffled on Model.constant().  I feel like the most recent commit improves usage, replacing args like `.{domain_sort.inner, range_sort.inner}` for an array constant with `.{domain_sort, range_sort}`.

Rexicon226 reviewed (approved):
I think it looks good enough. I have a few minor tweak I'd make, but I can do those later, no need to bother.

Rexicon226 commented:
Huge thanks for the help! I am going to implement a few more examples now, to get a feel for it. And then I'll merge!

archaistvolts commented:
For sure.  Thank you for this repo!  Looking forward to your follow up changes.  I've been working on another PR and wanted to base it on them.  Let me know if there is anything else I can do to make merging this easier such as squashing all these commits.

Rexicon226 commented:
Nope, nothing to do. I just need time to experiment a bit further. Excited to see those other changes!
